dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hdfs/ScanContext.java

```diff
@@ -61,30 +76,70 @@ final class ScanContext implements Closeable {
   @GuardedBy("csvPrinter")
   private long numDirsWalked = 0L;
 
-  ScanContext(DistributedFileSystem dfs, @WillClose Writer outputSink) throws IOException {
+  /**
+   * The only constructor.
+   *
+   * @param execManager the execution manager used to recursively perform the scan
+   * @param dfs the HDFS (parts of which are) going to be scanned
+   * @param outputSink where the results of the scan are written
+   * @throws IOException if something is wrong with the outputSink
+   */
+  ScanContext(ExecutorManager execManager, DistributedFileSystem dfs, @WillClose Writer outputSink)
+      throws IOException {
+    this.execManager = execManager;
     this.dfs = dfs;
     this.dfsClient = dfs.getClient();
     this.csvPrinter = AbstractTask.FORMAT.withHeader(HdfsFormat.Header.class).print(outputSink);
     this.instantScanBegin = Instant.now();
   }
 
+  /** Flush/close the output sink */
   @Override
   public void close() throws IOException {
     csvPrinter.close();
   }
 
+  /** Virtually a private method - should be referenced only by SingleDirScanJob.call */
   FileStatus[] listDirectory(FileStatus dir) throws IOException {
     Instant instantListBegin = Instant.now();
     FileStatus[] files = dfs.listStatus(dir.getPath());
     timeSpentInListStatus.add(Duration.between(instantListBegin, Instant.now()).toMillis());
     numFilesByListStatus.add(files.length);
+    numCallsToListStatus.increment();
     return files;
   }
 
-  /*
-   * CsvPrint the directory attributes of the specified dir
-   * and incrementally update the scan statistics (this is what the rest of the parameters are for)
+  /**
+   * The entry point of the ScanContext API - submits a (root) directory to be scanned and returns
+   * immediately. Multiple roots may be submitted. It does not make sense if one of them is
+   * (recursively) a subdir of another one. Use execManager.await() if you want to wait for the scan
+   * to complete.
+   *
+   * @param dir a hdfs directory to be scanned recursively. It should belong to the dfs file system.
    */
+  public void submitRootDirScanJob(FileStatus dir, ContentSummary contentSummary) {
+    if (contentSummary != null) {
+      synchronized (csvPrinter) {
+        hasContentSummary = true;
+        numFilesByContentSummary.add(contentSummary.getFileCount());
+        numDirsByContentSummary.add(contentSummary.getDirectoryCount());
+        startWalkDir(dir);
```

Comment:
Could you clarify why this `startWalkDir(dir);` call is under `synchronized` and the call below is not?

Reply:
If you look into startWalkDir's body, you'll see that all of it except for the last line is in `synchronized (csvPrinter) {`
I did this as a micro-optimization, but now I see it results in quite obfuscated code.
I got rid of the micro-optimization and simplified the code.